File: Peare/Resources/RT_VERSION/RT_VERSION.cs
using System;
using System.Text;
using System.Runtime.InteropServices;

namespace Peare
{
    public static class RT_VERSION
    {
        // The VS_HEADER structure is simplified to include only fields common to all blocks.
        private struct VS_HEADER
        {
            public ushort wLength;
            public ushort wValueLength;
        }

        public static string Get(byte[] data)
        {
            if (data == null || data.Length < 4)
                return "VERSIONINFO\n{\n}";

            int offset = 0;
            var sb = new StringBuilder();
            sb.AppendLine("VERSIONINFO");
            sb.AppendLine("{");

            int rootStart = offset;
            VS_HEADER rootHeader = ReadHeader(data, ref offset);

            bool isUnicode = IsUnicode(data, 6); // Offset 6 is where "VS_VERSION_INFO" should be
            if (isUnicode)
                offset += 2; // Unicode version has additional fields we're not interested in
            string rootKey = isUnicode ? ReadUnicodeZ(data, ref offset) : ReadAsciiZ(data, ref offset);
            Align4(ref offset);

            if (rootHeader.wValueLength > 0)
            {
                offset += rootHeader.wValueLength;
                Align4(ref offset);
            }

            // Start parsing the child blocks of the root
            ParseAndDump(data, ref offset, rootStart + rootHeader.wLength, sb, 1, isUnicode);

            sb.AppendLine("}");
            return sb.ToString();
        }

        private static void ParseAndDump(byte[] data, ref int offset, int parentEndOffset, StringBuilder sb, int indent, bool isUnicode)
        {
            while (offset < parentEndOffset && offset + 4 <= data.Length)
            {
                int currentBlockStart = offset;
                VS_HEADER header = ReadHeader(data, ref offset);

                if (header.wLength == 0 || currentBlockStart + header.wLength > parentEndOffset)
                    break; // Invalid block or one that exceeds the parent's boundary

                if (isUnicode)
                    offset += 2; // Skip two bytes when unicode

                string key = isUnicode ? ReadUnicodeZ(data, ref offset) : ReadAsciiZ(data, ref offset);
                Align4(ref offset); // Align after reading the key

                string indentStr = new string(' ', indent * 2);
                string value = null;

                // Handle the block's value
                if (header.wValueLength > 0)
                {
                    if (isUnicode)
                    {
                        if (indent == 3)
                        {
                            // ReadUnicodeZ reads the full value until the double null terminator and advances the offset
                            value = ReadUnicodeZ(data, ref offset);
                        }
                        else if (key == "Translation" && header.wValueLength >= 4)
                        {
                            // Special case for Translation (binary value)
                            ushort langID = BitConverter.ToUInt16(data, offset);
                            ushort codePage = BitConverter.ToUInt16(data, offset + 2);
                            value = $"{langID} {codePage}";
                            offset += header.wValueLength;
                        }
                    }
                    else // ASCII
                    {
                        if (indent == 3)
                        {
                            // ReadAsciiZ will find the null terminator (0x00) and advance the offset
                            value = ReadAsciiZ(data, ref offset);
                        }
                        else if (key == "Translation" && header.wValueLength >= 4)
                        {
                            // Special case for Translation (binary value)
                            ushort langID = BitConverter.ToUInt16(data, offset);
                            ushort codePage = BitConverter.ToUInt16(data, offset + 2);
                            value = $"{langID} {codePage}";
                            offset += header.wValueLength;
                        }
                    }

                    Align4(ref offset); // Align offset after reading (or skipping) the value
                }

                // Print the key and the value if present
                if (value != null)
                {
                    sb.AppendLine($"{indentStr}{key} = \"{Escape(value)}\"");
                }
                else
                {
                    sb.AppendLine($"{indentStr}{key}");
                }

                if (header.wValueLength == 0 && offset < currentBlockStart + header.wLength)
                {
                    sb.AppendLine($"{indentStr}{{");
                    // Recursively parse child blocks until the end of the current parent block
                    ParseAndDump(data, ref offset, currentBlockStart + header.wLength, sb, indent + 1, isUnicode);
                    sb.AppendLine($"{indentStr}}}");
                }

                offset = currentBlockStart + header.wLength;
                Align4(ref offset);
            }
        }

        private static VS_HEADER ReadHeader(byte[] data, ref int offset)
        {
            VS_HEADER header = new VS_HEADER
            {
                wLength = BitConverter.ToUInt16(data, offset),
                wValueLength = BitConverter.ToUInt16(data, offset + 2),
            };
            offset += 4;
            return header;
        }
        private static string ReadAsciiZ(byte[] data, ref int offset)
        {
            int start = offset;
            // Advance offset until a null terminator (0x00) is found or the end of data is reached.
            while (offset < data.Length && data[offset] != 0)
                offset++;

            int length = offset - start;
            string s = "";
            if (length > 0)
            {
                // Decode the ASCII string from the byte array.
                s = Encoding.ASCII.GetString(data, start, length);
            }

            // Skip the null terminator if it exists.
            if (offset < data.Length && data[offset] == 0)
                offset++;

            return s;
        }

        private static string ReadUnicodeZ(byte[] data, ref int offset)
        {
            int start = offset;
            // Advance offset by 2 bytes at a time (for UTF-16 characters) until a double null terminator (0x00 0x00) is found or the end of data is reached.
            while (offset + 1 < data.Length && !(data[offset] == 0 && data[offset + 1] == 0))
                offset += 2;

            int length = offset - start;
            string s = "";
            if (length > 0)
            {
                // Use the standard Unicode encoding (UTF-16 Little Endian).
                // If there are unmappable characters (e.g., problematic codepoints),
                // it will use the U+FFFD replacement character (the square box).
                s = Encoding.Unicode.GetString(data, start, length);

                // Now, replace the U+FFFD replacement character with a space.
                s = s.Replace('\uFFFD', ' ');
            }

            // Skip the double null terminator if it exists.
            if (offset + 1 < data.Length && data[offset] == 0 && data[offset + 1] == 0)
                offset += 2;

            return s;
        }

        private static bool IsUnicode(byte[] data, int offset)
        {
            // Unicode representation of "VS_VERSION"
            byte[] unicodeVersion = Encoding.Unicode.GetBytes("VS_VERSION");

            if (offset + unicodeVersion.Length > data.Length)
                return false;

            for (int i = 0; i < unicodeVersion.Length; i++)
            {
                if (data[offset + i] != unicodeVersion[i])
                    return false;
            }

            return true;
        }

        private static void Align4(ref int offset)
        {
            offset = (offset + 3) & ~3;
        }

        private static string Escape(string s)
        {
            return s.Replace("\\", "\\\\")
                     .Replace("\"", "\\\"")
                     .Replace("\t", "\\t")
                     .Replace("\n", "\\n")
                     .Replace("\r", "\\r");
        }
    }
}

File: src/decoder/resources/RT_VERSION/RT_VERSION.cpp
#include "RT_VERSION.h"

#include <QLatin1Char>
#include <QString>
#include <stdexcept>

namespace peare { namespace resources {
namespace {

// The VS_HEADER structure is simplified to include only fields common to all blocks.
struct VsHeader {
    quint16 length;
    quint16 valueLength;
};

quint16 readUInt16(const QByteArray& data, qsizetype offset)
{
    if (offset < 0 || offset + 2 > data.size())
        throw std::runtime_error("Attempted to read past end of version resource.");

    const unsigned char* p = reinterpret_cast<const unsigned char*>(data.constData() + offset);
    return quint16(p[0]) | (quint16(p[1]) << 8);
}

VsHeader readHeader(const QByteArray& data, qsizetype& offset)
{
    VsHeader header;
    header.length = readUInt16(data, offset);
    header.valueLength = readUInt16(data, offset + 2);
    offset += 4;
    return header;
}

void align4(qsizetype& offset)
{
    offset = (offset + 3) & ~qsizetype(3);
}

QString readAsciiZ(const QByteArray& data, qsizetype& offset)
{
    const qsizetype start = offset;
    // Advance offset until a null terminator (0x00) is found or the end of data is reached.
    while (offset < data.size() && data.at(offset) != '\0')
        ++offset;

    // Decode the ASCII string from the byte array.
    const QString value = QString::fromLatin1(data.constData() + start,
                                               int(offset - start));
    // Skip the null terminator if it exists.
    if (offset < data.size() && data.at(offset) == '\0')
        ++offset;
    return value;
}

QString readUnicodeZ(const QByteArray& data, qsizetype& offset)
{
    const qsizetype start = offset;
    // Advance offset by 2 bytes at a time (for UTF-16 characters) until a double null terminator (0x00 0x00) is found or the end of data is reached.
    while (offset + 1 < data.size()) {
        if (data.at(offset) == '\0' && data.at(offset + 1) == '\0')
            break;
        offset += 2;
    }

    QString value;
    // Use UTF-16 Little Endian. Unmappable replacement characters are converted to spaces.
    value.reserve(int((offset - start) / 2));
    for (qsizetype i = start; i + 1 < offset; i += 2) {
        const quint16 ch = readUInt16(data, i);
        value.append(QChar(ch == 0xFFFD ? quint16(' ') : ch));
    }

    // Skip the double null terminator if it exists.
    if (offset + 1 < data.size() && data.at(offset) == '\0' && data.at(offset + 1) == '\0')
        offset += 2;
    return value;
}

bool isUnicode(const QByteArray& data, qsizetype offset)
{
    // Unicode representation of "VS_VERSION".
    static const char marker[] = "VS_VERSION";
    const qsizetype count = qsizetype(sizeof(marker) - 1);
    if (offset + count * 2 > data.size())
        return false;

    for (qsizetype i = 0; i < count; ++i) {
        if (readUInt16(data, offset + i * 2) != quint16(static_cast<unsigned char>(marker[i])))
            return false;
    }
    return true;
}

QString escape(QString value)
{
    value.replace(QStringLiteral("\\"), QStringLiteral("\\\\"));
    value.replace(QStringLiteral("\""), QStringLiteral("\\\""));
    value.replace(QLatin1Char('\t'), QStringLiteral("\\t"));
    value.replace(QLatin1Char('\n'), QStringLiteral("\\n"));
    value.replace(QLatin1Char('\r'), QStringLiteral("\\r"));
    return value;
}

void parseAndDump(const QByteArray& data,
                  qsizetype& offset,
                  qsizetype parentEnd,
                  QString& output,
                  int indent,
                  bool unicode)
{
    const qsizetype safeParentEnd = qMin(parentEnd, qsizetype(data.size()));

    while (offset < safeParentEnd && offset + 4 <= data.size()) {
        const qsizetype blockStart = offset;
        const VsHeader header = readHeader(data, offset);
        const qsizetype blockEnd = blockStart + header.length;

        if (header.length == 0 || blockEnd > safeParentEnd)
            break; // Invalid block or one that exceeds the parent's boundary

        if (unicode)
            offset += 2; // Skip two bytes when unicode

        const QString key = unicode ? readUnicodeZ(data, offset)
                                    : readAsciiZ(data, offset);
        align4(offset); // Align after reading the key

        const QString indentation(indent * 2, QLatin1Char(' '));
        QString value;
        bool hasValue = false;

        // Handle the block's value.
        if (header.valueLength > 0) {
            if (indent == 3) {
                // Read the complete null-terminated string value.
                value = unicode ? readUnicodeZ(data, offset)
                                : readAsciiZ(data, offset);
                hasValue = true;
            } else if (key == QStringLiteral("Translation") &&
                       header.valueLength >= 4 && offset + 4 <= data.size()) {
                // Special case for Translation (binary value).
                const quint16 languageId = readUInt16(data, offset);
                const quint16 codePage = readUInt16(data, offset + 2);
                value = QStringLiteral("%1 %2").arg(languageId).arg(codePage);
                offset += header.valueLength;
                hasValue = true;
            }
            align4(offset); // Align offset after reading (or skipping) the value
        }

        // Print the key and the value if present.
        if (hasValue) {
            output += indentation;
            output += key;
            output += QStringLiteral(" = \"");
            output += escape(value);
            output += QStringLiteral("\"\n");
        } else {
            output += indentation;
            output += key;
            output += QLatin1Char('\n');
        }

        if (header.valueLength == 0 && offset < blockEnd) {
            output += indentation + QStringLiteral("{\n");
            // Recursively parse child blocks until the end of the current parent block.
            parseAndDump(data, offset, blockEnd, output, indent + 1, unicode);
            output += indentation + QStringLiteral("}\n");
        }

        offset = blockEnd;
        align4(offset);
    }
}

} // namespace

QString RT_VERSION::Get(const QByteArray& data)
{
    if (data.size() < 4)
        return QStringLiteral("VERSIONINFO\n{\n}");

    qsizetype offset = 0;
    QString output = QStringLiteral("VERSIONINFO\n{\n");

    const qsizetype rootStart = offset;
    const VsHeader rootHeader = readHeader(data, offset);
    const bool unicode = isUnicode(data, 6); // Offset 6 is where "VS_VERSION_INFO" should be

    if (unicode)
        offset += 2; // Unicode version has additional fields we are not interested in

    const QString rootKey = unicode ? readUnicodeZ(data, offset)
                                    : readAsciiZ(data, offset);
    Q_UNUSED(rootKey);
    align4(offset);

    if (rootHeader.valueLength > 0) {
        offset += rootHeader.valueLength;
        align4(offset);
    }

    // Start parsing the child blocks of the root.
    parseAndDump(data,
                 offset,
                 rootStart + rootHeader.length,
                 output,
                 1,
                 unicode);

    output += QStringLiteral("}\n");
    return output;
}

ResourcePreview RT_VERSION::preview(const ResourceEntry& entry)
{
    ResourcePreview preview;
    try {
        preview.text = Get(entry.data);
    } catch (const std::exception& error) {
        preview.error = QString::fromUtf8(error.what());
    }
    return preview;
}

} } // namespace peare::resources
